src/gui/windowsshutdownmanager.cpp

```diff
@@ -0,0 +1,97 @@
+/*
+ * Bittorrent Client using Qt4 and libtorrent.
+ * Copyright (C) 2014
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include <QDebug>
+#include <QString>
+#include <QApplication>
+#include <QSessionManager>
+
+#include <windows.h>
+#include "windowsshutdownmanager.h"
+
+typedef BOOL (WINAPI *PSHUTDOWNBRCREATE)(HWND, LPCWSTR);
+typedef BOOL (WINAPI *PSHUTDOWNBRDESTROY)(HWND);
+
+// this sets the shutdown/logoff blocking reason where supported (Vista or above)
+WindowsShutdownManager::WindowsShutdownManager(MainWindow *window) :
+    exitString(tr("Please wait while qBittorrent is exiting...\nBe patient, it could take a while...")),
+    window(window)
+{
+    // function pointer to ::ShutdownBlockReasonCreate function
+    // this is also used, in some way, to detect windows version
+    PSHUTDOWNBRCREATE shutdownBRCreate = (PSHUTDOWNBRCREATE)::GetProcAddress(::GetModuleHandle(TEXT("User32.dll")), "ShutdownBlockReasonCreate");
+  
+    if (shutdownBRCreate == NULL) {
+        qDebug() << "GetProcAddress for ShutdownBlockReasonCreate failed (this is fine on XP)";
+        
+        // this makes qBittorrent the first application to be requested to close
+        ::SetProcessShutdownParameters(0x3ff, SHUTDOWN_NORETRY);
+          
+        // on XP we need these connections in order to correctly manage the cancel button on the log-off dialog 
```

Comment:
Can you give a screenshot of this situation? I cannot understand on which dialog you are refering to.